File: verify_final.py
import asyncio
from playwright.async_api import async_playwright

async def run():
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True)
        page = await browser.new_page()

        # Log console messages
        page.on("console", lambda msg: print(f"CONSOLE: {msg.text}"))

        print("Loading page...")
        await page.goto("http://localhost:8080/editor.html", wait_until="networkidle")

        # Wait for AssetBrowser to be initialized
        await page.wait_for_function("window.assetBrowser !== undefined")
        print("AssetBrowser initialized")

        # 1. Create a folder via JS
        await page.evaluate("assetBrowser.createNewFolder('')")
        # Wait for prompt and fill it
        await page.wait_for_selector(".dialog-window")
        await page.fill(".dialog-input", "TestFolder")
        await page.click("button:has-text('Aceptar')")
        print("Folder 'TestFolder' created")

        # 2. Create an asset inside that folder
        await page.evaluate("assetBrowser.createNewAsset('ceanim', 'TestFolder')")
        await page.wait_for_selector(".dialog-window")
        await page.fill(".dialog-input", "InnerAnim")
        await page.click("button:has-text('Aceptar')")
        print("InnerAnim.ceanim created in TestFolder")

        # 3. Verify existence
        exists = await page.evaluate("FS.exists('TestFolder/InnerAnim.ceanim')")
        print(f"File TestFolder/InnerAnim.ceanim exists: {exists}")

        # 4. Test drag-over class behavior
        await page.evaluate("""() => {
            const el = document.getElementById('asset-browser');
            el.dispatchEvent(new DragEvent('dragenter', { bubbles: true }));
        }""")
        has_class = await page.evaluate("document.getElementById('asset-browser').classList.contains('drag-over')")
        print(f"Browser has drag-over class: {has_class}")

        await page.evaluate("""() => {
            const el = document.getElementById('asset-browser');
            el.dispatchEvent(new DragEvent('dragleave', { bubbles: true }));
        }""")
        has_class = await page.evaluate("document.getElementById('asset-browser').classList.contains('drag-over')")
        print(f"Browser has drag-over class after dragleave: {has_class}")

        # 5. Check Move-to-self bug
        # We need a file to move
        await page.evaluate("assetBrowser.createNewAsset('ces', '')")
        await page.wait_for_selector(".dialog-window")
        await page.fill(".dialog-input", "MoveTest")
        await page.click("button:has-text('Aceptar')")

        # Mock a drop event that moves MoveTest.ces to its own folder (root)
        await page.evaluate("""() => {
            const dt = new DataTransfer();
            dt.setData('text/plain', 'MoveTest.ces');
            const dropEvent = new DragEvent('drop', {
                bubbles: true,
                cancelable: true,
                dataTransfer: dt
            });
            document.getElementById('asset-browser').dispatchEvent(dropEvent);
        }""")

        exists_after_move = await page.evaluate("FS.exists('MoveTest.ces')")
        print(f"File MoveTest.ces exists after move-to-self: {exists_after_move}")

        await page.screenshot(path="/home/jules/verification/final_state.png")
        await browser.close()

asyncio.run(run())

File: verify_final.js
const { chromium } = require('playwright');

(async () => {
  const browser = await chromium.launch();
  const page = await browser.newPage();

  await page.addInitScript(() => {
    window.showDirectoryPicker = null;
  });

  await page.goto('http://localhost:8080/editor.html');
  await page.waitForFunction(() => window.editorInitialized === true, { timeout: 15000 });

  await page.evaluate(async () => {
    let m = window.SceneManager.currentScene.getAllMaterias().find(m => m.name === 'TestObject');
    if (!m) {
        const { createBaseMateria } = await import('./js/editor/MateriaFactory.js');
        m = createBaseMateria('TestObject');
        const { Transform, Gyzmo, SpriteRenderer } = await import('./js/engine/Components.js');
        const t = m.getComponent(Transform);
        t.x = 0; t.y = 50;

        const g = new Gyzmo(m);
        g.layers = [{ x: 0, y: 0, width: 100, height: 100, color: '#00ff00', name: 'Area UP' }];
        m.addComponent(g);

        // Add a sprite to verify upright orientation
        const sr = new SpriteRenderer(m);
        // Using a placeholder image or engine logo
        sr.sprite = new Image();
        sr.sprite.src = 'image/Logo_C.png';
        m.addComponent(sr);

        window.SceneManager.currentScene.addMateria(m);
        window.updateScene();
    }
    window.selectMateria(m.id);
    window.updateScene();
  });

  await page.waitForTimeout(2000); // Wait for logo to load
  await page.screenshot({ path: '/home/jules/verification/final_check_v4.png' });

  await browser.close();
})();
